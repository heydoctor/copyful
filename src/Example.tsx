import React from "react";
import { CopyfulProvider, getCopySomehow } from "./configureCopy";
import HeaderSection from "./HeaderSection";
import BodySection from "./BodySection";

function App() {
  return (
    <CopyfulProvider copy={getCopySomehow("en-us")}>
      <HeaderSection />{/* Example Using Hooks */}
      <BodySection />{/* Example Using HOC */}
    </CopyfulProvider>
  );
}

export default App;
