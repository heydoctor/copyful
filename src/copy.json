{
  "en-us": {
    "header": {
      "title": "Copyful",
      "subtitle": "Copy management shouldn't get in the way of code.",
      "nested": {
        "dynamicValues": "Yes, even dynamic values, like this number: {someNumber} and this string: {someString}."
      }
    }
  },
  "1337": {
    "header": {
      "title": "c0pyful",
      "subtitle": "c0py m4n463m3n7 5h0uldn'7 637 1n 7h3 w4y 0f c0d3.",
      "nested": {
        "dynamicValues": "y35, 3v3n dyn4m1c v4lu35, l1k3 7h15 numb3r: {someNumber} 4nd 7h15 57r1n6: {someString}."
      }
    }
  }
}