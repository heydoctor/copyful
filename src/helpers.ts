export const getInterpolatedCopy = <TCopy, TContext>(
  copy: TCopy,
  context: TContext
) => {
  let newCopy = { ...copy };
  findInterpolations(newCopy, context);
  return newCopy;
};

const findInterpolations = (copy: any, context: any) => {
  Object.keys(copy).forEach((key) => {
    const currItem = copy[key];
    if (typeof currItem === 'object') {
      findInterpolations(currItem, context);
    } else if (typeof currItem === 'string') {
      copy[key] = interpolateCopy(currItem, context);
    }
  });
};

const interpolateCopy = (str: string, context: any) => {
  const delimeterMatcher = /{\S+}/gm;
  const variables = (str.match(delimeterMatcher) || []).map((dV: string) =>
    dV.substring(1, dV.length - 1)
  );
  if (variables.length) {
    let interpolatedValue = str;
    variables.forEach((v) => {
      const target = `{${v}}`;
      const value = context[v] || target;
      interpolatedValue = interpolatedValue.replace(target, value);
    });
    return interpolatedValue;
  }
  return str;
};
