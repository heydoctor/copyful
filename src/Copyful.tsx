import * as React from 'react';
import { getInterpolatedCopy } from './helpers';
export type InterpolationValues = { [key: string]: string | number };

export const createCopyful = <TCopy,>(defaultCopy: TCopy) => {
  const Context = React.createContext(defaultCopy);

  const CopyfulProvider: (props: {
    copy?: TCopy;
    children: any;
  }) => JSX.Element = ({
    copy = defaultCopy,
    children,
  }: {
    copy?: TCopy;
    children: React.ComponentType;
  }) => {
    const value = copy || defaultCopy;
    return <Context.Provider value={value}>{children}</Context.Provider>;
  };

  const useCopy = (context: InterpolationValues = {}): TCopy => {
    const copy = React.useContext(Context);
    const interpolatedCopy = context
      ? getInterpolatedCopy(copy, context)
      : copy;
    return React.useMemo(() => interpolatedCopy, [interpolatedCopy]);
  };

  const withCopy = (Component: any) => {
    const WithCopy = React.forwardRef(function WithCopy(props: any, ref: any) {
      const { innerRef, ...rest } = props;
      const copy = useCopy() || defaultCopy;
      return <Component copy={copy} ref={innerRef || ref} {...rest} />;
    });

    return WithCopy;
  };

  return {
    CopyfulProvider,
    useCopy,
    withCopy,
  };
};

export default createCopyful;
