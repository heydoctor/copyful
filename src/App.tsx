import React from "react";
import { createCopyful, interpolateString } from "./Copyful";

const staticCopy = {
  "en-us": {
    header: {
      title: "Copyful",
      subtitle: "Copy management shouldn't get in the way of code.",
      dynamicValues:
        "Yes, even dynamic values, like this number: {0} and this string: {1}.",
    },
  },
  "1337": {
    header: {
      title: "c0pyful",
      subtitle: "c0py m4n463m3n7 5h0uldn'7 637 1n 7h3 w4y 0f c0d3.",
      dynamicValues:
        "y35, 3v3n dyn4m1c v4lu35, l1k3 7h15 numb3r: {0} 4nd 7h15 57r1n6: {1}.",
    },
  },
};

const { CopyfulProvider, useCopy } = createCopyful(staticCopy["1337"]);

function App() {
  return (
    <CopyfulProvider value={staticCopy["en-us"]}>
      <SomePage />
    </CopyfulProvider>
  );
}

function SomePage() {
  const copy = useCopy();
  const { title, subtitle, dynamicValues } = copy.header;
  const interpolated = interpolateString(dynamicValues, [1, '"Hello World"']);
  return (
    <div className="App">
      <header className="App-header">
        <h1>{title}</h1>
        <h2>{subtitle}</h2>
        <p>{interpolated}</p>
      </header>
    </div>
  );
}

export default App;
