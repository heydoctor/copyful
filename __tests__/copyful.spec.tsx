import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { createCopyful } from '../src';
import { copy } from '../example/copy';
import { getCopySomehow, Locales } from '../example/configureCopy';

const setup = (copyOverrideLocale: Locales = 'en-us') => {
  const { CopyfulProvider, useCopy } = createCopyful(getCopySomehow(copy, copyOverrideLocale));

  function TextComponent() {
    const { title, dynamicValues } = useCopy('header', {
      someNumber: 4,
      someString: 'heydoc',
    });

    return (
      <>
        <span>{title}</span>
        <span>{dynamicValues}</span>
      </>
    );
  }

  return render(
    <CopyfulProvider>
      <TextComponent />
    </CopyfulProvider>
  );
};

afterEach(cleanup);

describe('useCopy', () => {
  const { asFragment } = setup('en-us');

  it('returns strings and interpolates values with provided context', () => {
    expect(asFragment()).toMatchSnapshot();
  });
});
