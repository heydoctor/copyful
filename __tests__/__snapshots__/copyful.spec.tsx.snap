// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyfulProvider supports overriding default copy 1`] = `
<DocumentFragment>
  <span>
    c0pyful
  </span>
  <span>
    y35, 3v3n dyn4m1c v4lu35, l1k3 7h15 numb3r: 4 4nd 7h15 57r1n6: heydoc.
  </span>
</DocumentFragment>
`;

exports[`useCopy returns strings and interpolates values with provided context 1`] = `
<DocumentFragment>
  <span>
    Copyful
  </span>
  <span>
    Yes, even dynamic values, like this number: 4 and this string: heydoc.
  </span>
</DocumentFragment>
`;
